lib/src/stream_extensions.dart

```diff
@@ -54,4 +54,17 @@ extension StreamExtensions<T> on Stream<T> {
     });
     return completer.future;
   }
+
+  /// Returns a view of this stream that eagerly buffers all events until
+  /// `listen()` is called.
+  Stream<T> buffer() {
+    late StreamSubscription<T> subscription;
+    var controller = StreamController<T>(
+        onPause: () => subscription.pause(),
+        onResume: () => subscription.resume(),
+        onCancel: () => subscription.cancel());
+    subscription = listen(controller.add,
+        onError: controller.addError, onDone: controller.close);
+    return controller.stream;
```

Comment:
Consider avoiding `late` and needing wrappers by setting callbacks after creation:
```dart
  var controller = StreamController<T>.sync();
  var subscription = this.listen(controller.add, controller.addError, controller.close);
  controller
    ..onPause = subscription.pause
    ..onResume = subscription.resume
    ..onCancel = subscription.cancel;
  return controller.stream;
```

With `maxBuffer`/`weight` that becomes:
```dart
  if (maxBuffer != null) RangeError.checkNotNegative(maxBuffer, "maxBuffer");
  var controller = StreamController<T>.sync();
  var subscription = this.listen(null, controller.addError, controller.close); 
  if (maxBuffer != null) {
    var bufferedEvents = 0;
    subscription.onData((T value) {
      controller.add(value);
      bufferedEvents += weight?.call(value) ?? 1;
      if (bufferedEvents >= maxBuffer) subscription.pause();
    });
  } else {
    subscription.onData(controller.add);
  }
  controller.onListen = () {
    subscription.onData(controller.add);
    subscription.resume();
  };
  controller
    ..onPause = subscription.pause
    ..onResume = subscription.resume
    ..onCancel = subscription.cancel;
  return controller.stream;
```